README.md
<!-- <img src=".github/logo.png" /> -->

# omg.js
[![](https://travis-ci.org/PengJiyuan/omg.svg?branch=master)](https://travis-ci.org/PengJiyuan/omg)
[![][npm-image]][npm-url]
[![][downloads-image]][downloads-url]

Oh my goodness! Awesome Canvas Render Library. [Demo](https://omg.js.org)
<div align=center>
  <img src=".github/demo1.gif" />
</div>

<div align=center>
  <img src=".github/demo2.gif" />
</div>

<video src=".github/demo.mov">

## Install

**NPM**    

`npm install omg.js --save`

```javascript
const OMG = require('omg.js');
console.log(OMG); // ...
```

**DOWNLOAD**

[Source Code](dist/omg.min.js)

```html
<script src="path/omg.min.js"></script>
<script>
  console.log(OMG);// ...
</script>
```

## Example

```javascript
import omg from 'omg.js';

const stage = omg({
  element: document.getElementById('canvas'),
  width: 500,
  height: 500,
  enableGlobalTranslate: true, // 开启全局拖拽
  enableGlobalScale: true, // 开启全局缩放
  position: 'absolute', // canvas.style.position
  images: [], // 提前加载的图片列表
  prepareImage: true, // 提前加载图片
  // or
  prepareImage: () => {
    hideLoading();
  }
});

stage.init();

const rect = stage.graphs.rectangle({
  x: 120,
  y: 120,
  width: 200,
  height: 200,
  color: '#'+(~~(Math.random()*(1<<24))).toString(16)
}).on('mousedown', function() {
  console.log('click rect2');
}).on('mouseenter', function() {
  rect.color = '#'+(~~(Math.random()*(1<<24))).toString(16);
  stage.redraw();
}).on('mouseleave', function() {
  rect.color = '#'+(~~(Math.random()*(1<<24))).toString(16);
  stage.redraw();
}).on('dragin', function() {
  console.log('drag in rect2');
  rect.color = '#ffffff';
  stage.redraw();
}).on('dragout', function() {
  console.log('drag out rect2');
  rect.color = '#'+(~~(Math.random()*(1<<24))).toString(16);
  stage.redraw();
}).on('drop', function() {
  console.log('you drop on the rect2!');
  rect.color = '#000';
  stage.redraw();
}).config({
  drag: true,
  changeIndex: true
});

stage.addChild(rect);

stage.show();
```

## Usage
```javascript
import omg from 'omg.js';

const stage = omg({
  element: document.getElementById('canvas'),
  width: 500,
  height: 500,
  enableGlobalTranslate: true, // enable global drag events.
  enableGlobalScale: true, // enable global scale events.
});

// Init
stage.init();

stage.getVersion(); // v x.x.x
```

### Graphs

**Rectangle**

```javascript
const rect = stage.graphs.rectangle({
  x: 0,
  y: 0,
  width: 110,
  height: 110,
  rotate: 45, // not work when define radius
  radius: {
    tl: 6, // top left
    tr: 6, // top right
    bl: 6, // bottom left
    br: 6 // bottom right
  },
  color: '#514022'
});
```

**Line & Polyline**
```javascript
const polyline = stage.graphs.line({
  matrix: [
    [10, 180],
    [40, 50],
    [80, 180],
    [90, 80],
    [110, 100],
    [140, 50],
    [260, 180]
  ]
});
```

**Polygon**
```javascript
const polygon = stage.graphs.polygon({
  matrix: [
    [310, 120],
    [360, 120],
    [348, 230],
    [250, 340],
    [146, 200]
  ],
  color: 'black',
  style: 'stroke',
  lineWidth: 4
});
```

**Image**

```javascript
const image = stage.graphs.image({
  x: 0,
  y: 0,
  width: 800,
  height: 500,
  src: './img/timg.jpg'
});

// For more detail, check canvas api.
const image3 = stage.graphs.image({
  x: 200,
  y: 200,
  width: 97,
  height: 110,
  sliceX: 5,
  sliceY: 0,
  sliceWidth: 97,
  sliceHeight: 110,
  src: './img/action.png'
});
```

**Text**

```javascript
const text = stage.graphs.text({
  x: 300,
  y: 40,
  width: 150,
  height: 40, // x, y, width, height指定了一个矩形， 文字就在其中
  paddingTop: 8, // 用于调整文字在矩形中的位置
  center: true,  // 文字居中
  background: {
    color: 'blue', // 背景颜色
    img: './img/text_bg.png' // 背景图片
  }, // 给文字添加背景颜色(矩形)
  font: 'italic bold 20px arial,sans-serif', // 文字样式
  text: 'Hello stage', // 文字内容
  color: '#fff', // 文字颜色
  style: 'fill' // fill -- 填充， stroke -- 描边
});
```

**Circle**

```javascript
const arc = stage.graphs.arc({
  x: 400,
  y: 400,
  radius: 30,
  color: 'rgba(255, 255, 255, 0.5)',
  style: 'fill' // fill -- 填充， stroke -- 描边
});
```

**Sector**

```javascript
const arcb = stage.graphs.arc({
  x: 130,
  y: 380,
  radius: 100,
  startAngle: 45,
  endAngle: 165,
  color: '#512854',
  style: 'fill'
});
```

### Hide Shapes
```javascript
const arcb = stage.graphs.arc({
  ...,
  hide: true
});

// or

arcb.hide = true;
```

### Expand Shapes

If the default shapes not enough, you can expand shapes simply.
Before stage init, just expand your shapes by function `extend`.


```javascript
const stage = omg({
  ...
});

// demo shape
const yourShape = function(settings, _this) {
  const draw = function() {
    const canvas = _this.canvas;
    const scale = _this.scale;

    // if you want trigger events(includes drag, scale, mousedown, mouseenter...), you must add this line.
    stage.ext.DefineScale.call(this, scale, 'moveX', 'moveY', 'matrix');

    const matrix = this.scaled_matrix;

    canvas.save();
    canvas.translate(this.scaled_moveX, this.scaled_moveY);
    canvas.beginPath();

    matrix.forEach((point, i) => {
      i === 0 ? canvas.moveTo(point[0], point[1]) : canvas.lineTo(point[0], point[1]);
    });
    canvas.lineTo(matrix[0][0], matrix[0][1]);
    
    canvas.fillStyle = this.color;
    canvas.fill();
    canvas.closePath();
    canvas.restore();
  };

  return Object.assign({}, stage.ext.display(settings, _this), {
    type: 'polygon',
    draw: draw,
    lineWidth: settings.lineWidth || 1,
    matrix: settings.matrix
  });
};

// Before init, extend your shapes.
stage.extend({
  yourShape: yourShape
});

stage.init();

// use your extend shape
const shape = stage.graphs.yourShape({
  ...settings
});

stage.addChild(shape);

stage.show();

```

### Add Event

#### global events

Global event (mousedown, mousemove) for whole omg.

```javascript
stage.mousedown(function(e) {
  console.log(stage.utils.getPos(e));
});

stage.mousemove(function(e) {
  console.log(stage.utils.getPos(e));
});
```

#### shape events

All pc event types supported:

* mousedown
* mouseup
* mouseenter
* mouseleave
* mousemove'
* drag
* dragend
* dragin
* dragout
* drop

All mobile event types supported:

* touchstart
* touchmove
* touchend
* tap

Support chain call.

```javascript
/*!
 * @cur: current shape.
 */
shape.on('mousedown', function( cur ) {
  console.log('you click rect');
}).on('mousemove', function( cur ) {
  console.log('you move!');
}).on('mouseleave', function( cur ) {
  console.log('you leave!');
}).drag(true).config(){...};
```

### config

```javascript
rect.config({
  zindex: 10,
  drag: true, // 允许形状(对象)被拖拽  
  changeIndex: true， // 在形状被选中的时候允许改变个形状展示的顺序
  fixed: true, // 免受globalTranslate, globalScale的影响
});
```

### Group
You can insert some shapes into a group, so that then can behave consistency.
```javascript
const group = stage.group({
  x: 100,
  y: 100,
  width: 200,
  height: 200,
  title: {
    title: {
      text: 'Group Name',
      fontSize: 14,
      paddingTop: 12,
      paddingLeft: 14
    }
  },
  /**
   * @param {Object} background - 可以给group指定背景，目前只支持颜色
   * @TODO: 支持图片背景
   * /
  background: {
    color: '#000',
    img: '/path/x.png' // TODO
  },
  /**
   * @param {Object} border - 除了指定背景外，也可以给group指定边框，目前只支持颜色
   * @TODO: 支持border线宽
   * /
  // 如果background和border都不指定的话，group只做一个范围概念，同样支持事件触发。
  border: {
    color: '#000',
    lineWidth: 2 // TODO
  },
  zindex: 10
}).on('mousedown', function() {
  console.log('you clicked group');
});
```

* **Function** `group.add()`

Add child element to group.
The child element's coordinates will be referenced to the group's starting point.

* **Function** `group.remove()`

Remove an child element from group.

1. remove([Array])  -  remove shapes list
2. remove([Function])  -  remove group's children filter by function.

* **Function** `group.updateAllChildsPosition()`

update all group's children's position.

### Add shapes to stage.

```javascript
stage.addChild(rect);
stage.addChild(line);
stage.addChild(group);

// or

stage.addChild([rect, line, arc1, text1]);
```

### Remove shapes from stage.

```javascript
stage.removeChild(rect);

stage.removeChild([rect, arc, line]);

stage.removeFirstChild();

stage.removeLastChild();

stage.removeAllChilds();
```


### Show
Draw and trigger events.

```javascript
stage.show();
```

If you call function addChild or removeChild and bind some new events or unbind some events, you shold reset events trigger by:
```javascript
stage.show()
```
or
```javascript
stage.draw();
stage._events.triggerEvents();
```

### Draw and Redraw
```javascript
stage.draw();
stage.redraw();
```

### Reset
Reset OMG to init status.

```javascript
stage.reset();
```

### Animation

#### Global Animation

```javascript
function go() {
  rect.x++;
  line.y = line.y + 2;
  arc.radius++;
}

stage.animate(go);
```

#### Shap Animation

[Demo](https://omg.js.org/animation.html)

```javascript
/**
 * @param: {keys | Object}   -- the end value of your tween
 * @param: {config | Object} -- the tween settings
 */
shape.animateTo({
  x: 100,
  y: 100,
  width: 200,
  height: 200
}, {
  duration: 1000, //default = 500
  delay: 500, // defalut = 0
  easing: 'bounceOut', // defalut = 'linear'
  onStart: function(keys) {
    /**
     * @param: keys
     * the values during you tween
     * same below
     */
    console.log(keys.x, keys.y, keys.width, keys.height);
  }, // defalut = undefined
  onUpdate: function(keys) {
    console.log(keys.x, keys.y, keys.width, keys.height);
  }, // defalut = undefined
  onFinish: function(keys) {
    console.log(keys.x, keys.y, keys.width, keys.height);
  }, // defalut = undefined
});
```

#### Easing Types
* linear
* quadIn
* quadOut
* quadInOut
* cubicIn
* cubicOut
* cubicInOut
* quartIn
* quartOut
* quartInOut
* quintIn
* quintOut
* quintInOut
* sineIn
* sineOut
* sineInOut
* bounceOut
* bounceIn
* bounceInOut

#### Clear Animation
```javascript
stage.clearAnimation();
```

#### After Finish Animation
```javascript
stage.finishAnimation = () => {
  console.log('all animations has finished!');
};
```

### Resize

#### resize(opt)
* opt.width {Function} -- width after resize
* opt.height {Function} -- height after resize
* opt.resize {Function} -- callback triggered after resize

```javascript
world.resize({
  width: () => document.body.clientWidth,
  height: () => document.body.clientHeight,
  // function update - update the canvas size
  resize: (update) => {
    update();
  }
});
```

### FPS

#### fpsOn

If you use some animations and want to know the fps, just call function `fpsOn` to get fps.

```javascript
stage.fpsOn(function(fps) {
  console.log(fps);
});
```

#### Fps off
```javascript
stage.fpsOff();
```

## [CHANGELOG](./.github/CHANGELOG.md)

## [MIT](./LICENSE)

[npm-image]: https://img.shields.io/npm/v/omg.js.svg
[npm-url]: https://npmjs.org/package/omg.js
[downloads-image]: https://img.shields.io/npm/dm/omg.js.svg
[downloads-url]: https://npmjs.org/package/omg.js


.github/CHANGELOG.md
### v1.0.0
* 添加矩形，线的绘制
* 已on的形式给对象添加事件
* 全局拖拽位移和缩放
* 对象的拖拽位移

### v1.1.0
* 优化事件触发器，新增`dragin` `dragout` `drop`事件， 用于对象之间的交互， 模拟原生的drag and drop。
* 新增对象之间的**两两交互**，和原生drag，drop类似，不过简化为了dragin, dragout, drop， 这三个事件可以完美解决两两交互, **注意**这三个属性是作为contaner的一方添加。
* 结合**外部drag，drop**，和页面内交互。OMG作为一个全局变量，包含了一切需要的数据，何以直接结合外部drag，drop操作objects数组，然后重绘。
* 优化事件触发的顺序(order)， 支持改变对象绘制的顺序。
* 对象添加属性链式调用。XX.on(..).on(..).config(...).drag(...)...
* 优化点在形状内的判断，添加形状旋转后的判断...

#### v1.1.1
* 添加scripts文件，新增release脚本，自动化更新版本，提交代码。
* 添加eslint， 和pre-commit脚本。

### v1.2.0
* config添加fixed, bg 。 `fixed:ture` --> globalTranslate对当前对象无效
* 暂时移除rotate
* 添加图片加载器
* 添加 drag,dragend事件， 用于对象在拖拽和拖拽结束的时候触发，drop事件的callback添加item回调， `on('drop', function(item) {console.log(item)})` item是被拖拽的对象。
* 添加图形 arc,text,image..
* 优化Event

### v1.3.0
* 由于之前的设定， 只有一个全局变量OMG，而所有的信息，包括全局位置信息等都储存在着一个全局变量之中，导致无法绘制多个canvas，他们会共享一个全局变量。所以改版，支持多canvas绘制。

#### v1.3.1
* 绘制图形增添zindex, 默认为0, 可以为负数。
* addChild可以直接添加数组。

#### v1.3.4
*  新增绘制坐标系(coord.js)，只需提供x，y轴数据，即可自己选取合适的区间，绘制坐标系。

#### v1.3.5
*  增加颜色函数，包括十六进制颜色转rgb颜色， rgb颜色转hsl颜色， 仿less的lighten和darken函数...

#### v3.0.0-beta.0
*  重构代码。
*  添加autoscale，解决Retina屏模糊问题。
*  Event callback添加参数self, 可以在callback中直接操作当前shape
*  添加不规则图形的绘制。
*  添加点在不规则图形内的判断。

#### v3.0.0-beta.1
* config配置的时候不能传入false，bugfix。
* 添加全局mousedown, mousemove事件配置。
* stage添加removeChild, removeFirstChild, removeLastChild, removeAllChilds方法.

#### v3.0.0-beta.2
* 添加动画接口，包括全局动画和指定图形的补间动画（支持各种曲线缓动）。
* 优化动画接口，全局动画和图形补间动画在一个动画队列中，保证多个动画的有序进行。

#### v3.0.0-beta.3
* 添加fps开关.

#### v3.0.0
* 添加全局缩放功能，而且缩放之后不影响事件的触发。
* 添加resize函数
* 优化代码结构，将判断点在区域内的函数抽离到单独的文件中

#### v3.1.0
* 添加group的概念，group也可以当作图形来理解，但是可以添加子图形。子图形的参照起点坐标将会是group的起点。
* group添加remove函数，用于移除group中的子图形。
* group添加子组件新增updated判断和forceUpdate选项。
* 添加clip。绘制动画的时候（比如线）支持动画绘制（从左向右展开）。
* resize事件优化，canvas支持改变style.position。
* group支持添加title绘制。
* onmouseenter, onmouseleave事件优化
* 图形Rectangle新增border-radius。
* 图形text添加图片背景选项。
* 修复获取鼠标坐标点异常的问题。

#### v3.1.1
* 修复group的title不存在时的bug.
* 调整group及子图形的zindex的计算方法。
* group.remove增加function类型的参数。`group.remove(g => g.type === 'polygon')`.
* 增加`group.updateAllChildsPosition()`, 用于调用更新group的所有子图形的坐标.

#### v3.1.2
* 当shape的属性为hide时，在触发事件的时候将之忽略。

#### v3.1.3
* `gruop.add()`的时候不对child进行排序。
* 在初始化的时候，新增对 初始缩放比例(deviceScale), 最大缩放比例(maxDeviceScale), 最小缩放比例(minDeviceScale)的操作。

#### v3.1.4
* 初始化参数prepareImage 增加 函数类型，会在加载完图片之后触发，可用于在加载完图片之后隐藏loading层。
* 修复animateTo的onUpdate的触发顺序。

#### v4.0.0-beta.0
* 添加flow。
* graph的绘制接口调整，不向下兼容。直接通过stage.shape(...)调整为stage.graphs.shape(...)。

#### v4.0.0
* 修复无法移除wheel事件的bug。
* 添加api `setGlobalProps`， 用于设置OMG全局状态比如enableGlobalTranslate, enableGlobalScale。
* Graph的config移除bg参数。
* Graph的fixed属性升级，设置了fixed之后，全局拖拽，全局缩放，自身拖拽全部对其无效。
* 添加页面重置api reset。
* graph line支持拖拽
* 添加finishAnimation接口。当omg中所有的动画结束后会执行这一函数。

## ToDo

* 给时间触发添加顺序(index) (Done)
* 全局拖拽位移和缩放 -- (Done)
* 对象之间的两两交互，模拟drag, drop事件, 新增dragin, dragout, drop事件。(Done)
* 外部拖拽与页面内对象交互，像原生的drag，drop. (Done) -- 见demo源代码
* 浏览器兼容性测试
* 添加基本，常见的图形的绘制（Done）
* 增加group组件，几个图形可以放到一个group中，作为一个整体来添加事件和操作。(Done)

src/core.js
/* @flow */

import './utils/polyfill';
import { version } from '../package.json';
import event from './event/index';
import { Color } from './utils/color';
import { ImageLoader } from './utils/imageLoader';
import { Tween } from './tween/index';
import autoscale from './utils/autoscale';
import * as utils from './utils/helpers';
import shapes from './shapes/index';
import group from './group/index';
import clip from './clip/index';
import * as ext from './extend/export';

export type configSettings = {
  deviceScale: ?number,
  minDeviceScale: ?number,
  maxDeviceScale: ?number,
  prepareImage: boolean | Function | void,
  element: HTMLCanvasElement,
  width: number,
  height: number,
  position: ?string,
  enableGlobalTranslate: ?boolean,
  enableGlobalScale: ?boolean,
  images: ?Array<string>
}

export class OMG {

  version: string;                       // OMG's current version
  isMobile: boolean;                     // Current device is mobile phone
  objects: Array<GraghShape>;            // All shapes list
  _objects: Array<GraghShape>;           // All shapes list's reverse list
  groupRecords: number;                  // For generating and recording the graphs' zindex in a group
  transX: number;                        // The number of global translate x
  transY: number;                        // The number of global translate y
  deviceScale: number;                   // Default scale rate
  minDeviceScale: number;                // Minimum scale rate
  maxDeviceScale: number;                // Maximum scale rate
  scale: number;                         // Current scale rate
  loader: Object;                        // Instance of imageLoader
  // Ture or a function can be define. Before render, will load images first.
  // If prepareImage is a function, will trigger after images loaded.
  prepareImage: boolean | Function | void;
  globalMousedown: Function | void;      // Global mousedown function.
  globalMousemove: Function | void;      // Global mousemove function.
  Tween: any;                            // Class Tween
  animationList: Array<any>;             // The List contains page's all animation instance.
  animating: boolean | void;             // Whether the page is animating
  animationId: number | void;            // Animation's id.
  cacheIdPool: Object;                   // An object contains animationId
  fpsFunc: Function | void;              // If define fpsFunc, can get real-time fps.
  fps: number;                           // Real-time fps.
  fpsCacheTime: number;                  // Used to cache timestamps which used to calculate fps.
  eventTypes: Array<string>;             // All supported event types list.
  mobileEventTypes: Array<string>;       // All supported mobile event types list.
  _event: Object;                        // Instance of class event.
  color: Color;                          // Class color.
  element: HTMLCanvasElement;            // Element canvas.
  canvas: CanvasRenderingContext2D;      // canvas.getContext2D()
  width: number;                         // canvas's with
  height: number;                        // canvas's height
  ext: Object;                           // Export functions for extends custom graphs.
  clip: Function;                        // Export clip function.
  enableGlobalScale: boolean | void;     // Enable global translate?
  enableGlobalTranslate: boolean | void; // Enable global scale?
  images: Array<string>;                 // The image list for preload.
  utils: Object;                         // Some helper functions
  shapes: Object;                        // All shapes function.
  // shapes
  group: Function;                        // Group instance
  graphs: {[graph_name: string]: Object}; // graphs contains all graphs' instance.

  constructor(config: configSettings) {

    this.version = version;

    this.isMobile = utils.isMobile();

    this.objects = [];

    this.groupRecords = 0,

    this.transX = 0;

    this.transY = 0;

    this.deviceScale = config.deviceScale || 1;

    this.minDeviceScale = config.minDeviceScale || 0.5 * this.deviceScale;

    this.maxDeviceScale = config.maxDeviceScale || 4 * this.deviceScale;

    this.scale = this.deviceScale;

    this.loader = new ImageLoader();

    this.prepareImage = config.prepareImage;

    this.globalMousedown = void(0);

    this.globalMousemove = void(0);

    this.Tween = Tween;

    this.animationList = [];

    this.animationId = 0;

    this.animating = false;

    this.cacheIdPool = {};

    this.fpsFunc = void(0);

    this.fps = 0;

    this.fpsCacheTime = 0;

    this.graphs = {};

    this.eventTypes = [
      'click',
      'mousedown',
      'mouseup',
      'mouseenter',
      'mouseleave',
      'mousemove',
      'drag',
      'dragend',
      'dragin',
      'dragout',
      'drop'
    ];

    this.mobileEventTypes = [
      'touchstart',
      'touchend',
      'touchmove',
      'tap',
      'pinch',
      'spread',
      'drag',
      'dragend',
      'dragin',
      'dragout',
      'drop'
    ];

    this._event = event(this, this.isMobile);

    this.color = new Color();

    this.element = config.element;

    this.canvas =  this.element.getContext('2d');

    this.width = config.width;

    this.height = config.height;

    autoscale([this.element], {
      width: this.width,
      height: this.height,
      position: config.position || 'relative'
    });

    /**
     * @description: For extend shapes.
     *               Export functions to define scale and drag events...
     */
    this.ext = ext;

    this.clip = clip;

    // enable global drag event.
    this.enableGlobalTranslate = config.enableGlobalTranslate || false;

    // enable global scale event.
    this.enableGlobalScale = config.enableGlobalScale || false;

    // init images
    this.images = config.images || [];

    this.utils = utils;

    this.shapes = shapes;

  }

  init() {
    for(let shape in this.shapes) {
      this.graphs[shape] = settings => {
        return this.shapes[shape](settings, this);
      };
    }

    this.group = function(settings) {
      return group(settings, this);
    };
  }

  reset() {
    this.transX = 0;
    this.transY = 0;
    this.scale = this.deviceScale;
    this.objects.filter(o => !o.parent).forEach(o => {
      o.moveX = 0;
      o.moveY = 0;
    });
    this.objects.filter(o => o.type === 'group').forEach(o => {
      o.updateAllChildsPosition();
    });
    this._objects = utils.reverse(this.objects);
    this.redraw();
  }

  extend(ext: Object): void {
    for (let key in ext) {
      this.shapes[key] = ext[key];
    }
  }

  // Confused flow.
  setGlobalProps(props: {[prop_name: string]: any}) {
    for(let key in props) {
      switch(key) {
        case 'enableGlobalTranslate':
          this.enableGlobalTranslate = props[key];
          break;
        case 'enableGlobalScale':
          this.enableGlobalScale = props[key];
          break;
        default:
          break;
      }
    }
    this._event.triggerEvents();
  }

  // Array<Object> | Object
  addChild(child: any) {
    // multi or single
    if(utils.isArr(child)) {
      this.objects = this.objects.concat(child);
    } else if(utils.isObj(child)) {
      this.objects.push(child);
    }
    this.objects.sort((a, b) => {
      return a.zindex - b.zindex;
    });
    // copy the reverse events array
    this._objects = utils.reverse(this.objects);
  }

  removeChild(child: Array<Object> | Object) {
    if(utils.isArr(child)) {
      this.objects = this.objects.filter(o => !~child.indexOf(o));
    } else {
      this.objects = this.objects.filter(o => o !== child);
    }
    this._objects = utils.reverse(this.objects);
  }

  removeFirstChild() {
    this.objects.pop();
    this._objects = utils.reverse(this.objects);
  }

  removeLastChild() {
    this.objects.shift();
    this._objects = utils.reverse(this.objects);
  }

  removeAllChilds() {
    this.objects = [];
    this._objects = [];
  }

  imgReady() {
    this.loader.addImg(this.images);
  }

  show() {
    const _this = this;
    // dirty, ready to remove
    if(this.prepareImage) {
      this.imgReady();
      this.loader.ready(() => {
        typeof this.prepareImage === 'function' && this.prepareImage();
        _this.draw();
        _this._event.triggerEvents();
      });
    } else {
      this.draw();
      this._event.triggerEvents();
    }
  }

  draw() {
    this.objects.forEach(item => {
      item.draw();
    });
  }

  redraw() {
    this.clear();
    this.canvas.save();
    this.canvas.translate(this.transX, this.transY);
    this.draw();
    this.canvas.restore();
  }

  clear() {
    this.canvas.clearRect(0, 0, this.width, this.height);
  }

  tick() {
    const func = () => {
      if(this.fpsFunc) {
        const now = Date.now();
        if(now - this.fpsCacheTime >= 1000) {
          this.fpsFunc && this.fpsFunc(this.fps);
          this.fps = 0;
          this.fpsCacheTime = now;
        } else {
          this.fps++;
        }
      }
      this.animationList.forEach((t, i) => {
        // if finished, remove it
        if(t.finished) {
          this.animationList.splice(i--, 1);
        } else if(t.update) {
          t.update();
        } else {
          t();
        }
      });
      this.redraw();
      if(this.animationList.length === 0 && this.animating) {
        this.animating = false;
        this.finishAnimation();
        cancelAnimationFrame(this.cacheIdPool[this.animationId]);
      } else {
        this.cacheIdPool[this.animationId] = requestAnimationFrame(func);
      }
    };
    if(this.animationList.length > 0 && !this.animating) {
      this.animating = true;
      this.animationId = Date.now();
      func();
    }
    return this.animationId;
  }

  finishAnimation() {}

  /**
   * @param {func | Function}
   * The func you get the fps and do something.
   *
   * eg.
   * stage.fpsOn(function(fps) {
   *   console.log(fps);
   * });
   */
  fpsOn(func: Function) {
    this.fpsFunc = func;
    this.fpsCacheTime = Date.now();
  }

  fpsOff() {
    this.fpsFunc = void(0);
    this.fps = 0;
  }

  // add an animation to animationList.
  animate(func: Function) {
    this._event.triggerEvents();
    this.animationList.push(func);
    this.tick();
  }

  // clear all animations, includes global animation and shape animations.
  clearAnimation() {
    this.animationList = [];
    this.animating = false;
    cancelAnimationFrame(this.cacheIdPool[this.animationId]);
  }

  // get current version
  getVersion() {
    return this.version;
  }

  // global mousedown event.
  mousedown(func: Function) {
    this.globalMousedown = func;
  }

  // global mousemove event
  mousemove(func: Function) {
    this.globalMousemove = func;
  }

  /**
   *
   * @param {Object} opt
   *
   * @param {Function} opt.width  - width after resize
   * @param {Function} opt.height - height after resize
   * @param {Function} opt.resize - callback triggered after resize
   */
  resize(opt: {
    width: () => number,
    height: () => number,
    resize: Function
  }) {
    const update = () => {
      this.width = opt.width();
      this.height = opt.height();
      autoscale([this.element], {
        width: this.width,
        height: this.height,
        position: 'absolute'
      });
      this.redraw();
    };
    if(!window.onresize) {
      utils.bind(window, 'resize', () => {
        if(opt.resize) {
          opt.resize(update);
        } else {
          update();
        }
      });
    }
  }

}
